meta-balena-common/recipes-support/resin-init/resin-init-flasher/resin-init-flasher

```diff
@@ -133,26 +171,171 @@ if [ -z "$internal_dev" ]; then
+    dd if=/dev/zero of="/dev/$internal_dev" bs=4M count=1
+    # Regardless of what the original image uses we always want GPT for secure boot + LUKS
+    # Though in practice MBR would work as well in most cases, it is not globally guaranteed
+    # and it is much harder to operate on due to the necessity of an extended partition
+    parted "/dev/$internal_dev" mktable gpt
+
+    inform "Flashing boot partition"
+    ORIGINAL_BOOT_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-boot)
+    ORIGINAL_BOOT_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$ORIGINAL_BOOT_PART_ID")
+    ORIGINAL_BOOT_START=$(get_part_start_by_number "$LOOP_DEVICE_NAME" "$ORIGINAL_BOOT_PART_ID")
+
+    parted "/dev/$internal_dev" unit MiB mkpart balena-efi "$ORIGINAL_BOOT_START" $["$ORIGINAL_BOOT_START" + "$ORIGINAL_BOOT_PART_SIZE"]
+
+    EFI_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-boot)
+
+    PART_PREFIX=""
+    if [ -e "/dev/${internal_dev}p${EFI_PART_ID}" ]; then
+        PART_PREFIX="p"
     fi
-    IMAGE_WRITTEN_BYTES=$(awk 'END{print $1}' /tmp/dd_progress_log)
-    let RATIO=$IMAGE_WRITTEN_BYTES*100/$IMAGE_FILE_SIZE || true
-    resin-device-progress --percentage $RATIO --state "Flashing balenaOS on internal media" || true
-    truncate -s 0 /tmp/dd_progress_log
-done
+
+    dd if="${LOOP_DEVICE}p${ORIGINAL_BOOT_PART_ID}" of="/dev/${internal_dev}${PART_PREFIX}${EFI_PART_ID}" bs=4M
+    FLASHED=$["$ORIGINAL_BOOT_PART_SIZE" * 1024 * 1024]
+
+    # Relabel former boot partition
+    fatlabel "/dev/${internal_dev}${PART_PREFIX}${EFI_PART_ID}" "balena-efi"
+
+    # Find parition IDs
+    ROOTA_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-rootA)
+    ROOTB_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-rootB)
+    STATE_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-state)
+    DATA_PART_ID=$(get_part_number_by_label "$LOOP_DEVICE_NAME" resin-data)
+
+    # Find partition sizes
+    BOOT_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$EFI_PART_ID")
+    ROOTA_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$ROOTA_PART_ID")
+    ROOTB_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$ROOTB_PART_ID")
+    STATE_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$STATE_PART_ID")
+    DATA_PART_SIZE=$(get_part_size_by_number "$LOOP_DEVICE_NAME" "$DATA_PART_ID")
+
+    # Find the beginning of the first partition
+    FIRST_PART_ID=$(echo -e "$ROOTA_PART_ID\n$ROOTB_PART_ID\n$STATE_PART_ID\n$DATA_PART_ID" | sort | head -n 1)
+    FIRST_PART_START=$(get_part_start_by_number "$LOOP_DEVICE_NAME" "$FIRST_PART_ID")
+
+    # Create new partitions with extra space for LUKS headers
+    # LUKS header size is 2MiB
+    LUKS_HEADER_SIZE=2
+
+    BOOT_PART_END=$["$FIRST_PART_START" + "$BOOT_PART_SIZE" + "$LUKS_HEADER_SIZE"]
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-boot "$FIRST_PART_START" "$BOOT_PART_END"
```

Comment:
All these partitions should be aligned - hence my comment of making sure that the output of the `get_part_size` functions is disk aligned.

Reply:
Added